tfhe/src/core_crypto/gpu/mod.rs

```diff
@@ -134,12 +141,15 @@ impl CudaStream {
     {
         assert!(dest.len() >= src.len());
         let size = dest.len() * std::mem::size_of::<T>();
-        cuda_memcpy_async_gpu_to_gpu(
-            dest.as_mut_c_ptr(),
-            src.as_c_ptr(),
-            size as u64,
-            self.as_c_ptr(),
-        );
+        // We check that src and dest are not empty to avoid invalid pointers
+        if !src.is_empty() && !dest.is_empty() {
+            cuda_memcpy_async_gpu_to_gpu(
+                dest.as_mut_c_ptr(),
+                src.as_c_ptr(),
+                size as u64,
```

Comment:
we know that dest.len() > = src.len() and we copy size bytes of data coming from dest (which is potentially longer than source), meaning we can go past the end of src ?

I think the logic is wrong-ish in other functions, what would be safe I believe would be:

either take the min of src.len() and dest.len() but it feels like it could be wrong then, or rather check that the src len is smaller than dest and copy the content from source using the source size ?

Reply:
Yes the logic was not good here: I'll check that dest.len() >= src.len() then use src size for the copy (and check that src_size is > 0)